### ISSUE TITLE
Make it possible for type checking to detect "The truth value of a DataFrame is ambiguous."

### ISSUE THREAD

**Issue by karolzlot** [pandas-dev/pandas-stubs#663]
I described the problem here:
https://github.com/microsoft/pyright/issues/5024
and also here:
https://github.com/microsoft/pylance-release/issues/4289

**Comment by Dr-Irv** [comment]
I think the solution here is that `DataFrame.__eq__()` needs to have a stub in `core/frame.pyi` returning a `DataFrame`

**Comment by gandhis1** [comment]
> I think the solution here is that `DataFrame.__eq__()` needs to have a stub in `core/frame.pyi` returning a `DataFrame`

I'm not sure I follow what that will accomplish. This error is triggered when someone does the equivalent of casting a `DataFrame` to a `bool` - which calls `__bool__`, no? 

Even `pandas` itself has it annotated the way I suggested:

```
    @final
    def __nonzero__(self) -> NoReturn:
        raise ValueError(
            f"The truth value of a {type(self).__name__} is ambiguous. "
            "Use a.empty, a.bool(), a.item(), a.any() or a.all()."
        )

    __bool__ = __nonzero__
```

**Comment by Dr-Irv** [comment]
> I'm not sure I follow what that will accomplish. This error is triggered when someone does the equivalent of casting a `DataFrame` to a `bool` - which calls `__bool__`, no?

You're right, and actually `DataFrame.__eq__()` is return a `DataFrame` already.

So I tried with a simple class, and it didn't work with pyright.  See https://github.com/microsoft/pyright/issues/5039

**Comment by Dr-Irv** [comment]
pyright fixed the issue, in 1.1.306 so we should define `__bool__()` for `Series` and `DataFrame` as `NoReturn`, and add some tests.
